fakeserver/main.go

```diff
@@ -158,6 +212,54 @@ func tokenAuthStreamInterceptor(expectedToken string) grpc.StreamServerIntercept
 	}
 }
 
+// MarshalLogObject implements the zapcore.ObjectMarshaler interface
+func (r *RawMessageWrapper) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	var data map[string]interface{}
+	if err := json.Unmarshal(r.Raw, &data); err != nil {
+		return err
+	}
+	// Use the encoder to add the fields to the log
+	for k, v := range data {
+		enc.AddReflected(k, v)
+	}
+	return nil
+}
+
+func recordStreamedLog(log *pb.Log, logger zap.Logger) error {
+	// Decode the JSON-encoded string into a LogEntry struct
+	var logEntry LogEntry
+	if err := json.Unmarshal([]byte(log.JsonMessage), &logEntry); err != nil {
+		logger.Error("Error decoding JSON:", zap.Error(err))
+	}
+
+	// Convert the level from string to zapcore.Level
+	var level zapcore.Level
+	if err := level.UnmarshalText([]byte(logEntry.Level)); err != nil {
+		logger.Error("Error converting log level", zap.Error(err))
+	}
+
+	// Declare a variable of type json.RawMessage to hold the decoded data
+	var rawMessage json.RawMessage
+
+	// Decode the JSON-encoded string into the rawMessage
+	err := json.Unmarshal([]byte(log.JsonMessage), &rawMessage)
+	if err != nil {
+		logger.Error("Error decoding JSON:", zap.Error(err))
+		return err
+	}
+
+	// Wrap the rawMessage in RawMessageWrapper
+	wrappedRawMessage := &RawMessageWrapper{Raw: rawMessage}
+
+	if ce := logger.Check(level, "Received log from cloud-operator"); ce != nil {
+		ce.Write(
+			//zap.String("cluster_id", ...),
+			zap.Object("message", wrappedRawMessage),
+		)
+	}
```

Comment:
As I commented above, I think you can avoid doing this wrapping by just having `LogEntry` implement `ObjectMarshaler` directly.
```suggestion
	if ce := logger.Check(level, "Received log from cloud-operator"); ce != nil {
		ce.Write(
			//zap.String("cluster_id", ...),
			zap.Object("message", logEntry),
		)
	}
```